lib/THC/THCAtomics.cuh

```diff
@@ -102,9 +102,16 @@ static inline  __device__ void atomicAdd(half *address, half val) {
 
   do {
     assumed = old;
+#if CUDA_VERSION < 9000
     half hsum;
     hsum.x = (size_t)address & 2 ? (old >> 16) : (old & 0xffff);
     hsum = THCNumerics<half>::add(hsum, val);
+#else
+    __half_raw hsum;
+    hsum.x = (size_t)address & 2 ? (old >> 16) : (old & 0xffff);
+    half tmpres = THCNumerics<half>::add(hsum, val);
+    hsum.x = *(unsigned short*)&tmpres;
```

Comment:
while this probably works, technically it's UB (because of the strict aliasing rule). It would be better to change it to `hsum.x = reinterpret_cast<unsigned short>(tmpres);`. Same for all casts below